KEYCLOAK-19297 Use real 'external' client object id to store Authenti…

lgraf opened the pull request:
[KEYCLOAK-19297](https://issues.redhat.com/browse/KEYCLOAK-19297) Use real 'external' client object id to store AuthentcatedClientSession in UserSession object, so that the client session can be looked up in further requests.

hmlnarik commented:
@lgraf Thank you for the PR.

Could you please add a test for this functionality?

lgraf commented:
@hmlnarik Thank you for looking into this. I will try to add a test next week.

Commit:
KEYCLOAK-19297 Use real 'external' client object id to store AuthenticatedClientSession in UserSession object, so that the client session can be looked by the client object id in further requests.

lgraf commented:
@hmlnarik I added a model test that use a `ClientStorageProvider` that persists/load a `UserSessionModel` that contains a local and an external `AuthenticatedClientSessionModel`. With the test the bug can be reproduced (just revert the commit that contains the fix).

I introduced a new maven profile in the model tests so that the new `HardcodedClientStorage` model attribute  is loaded to allow that a `ClientStorageProvider` can be used during the test. Not quite sure if it is the only approach, though.

martin-kanis reviewed (dismissed):
@lgraf Thank you for the PR and sorry for the late review. It looks good to me. I put one comment to the test regarding persisting online session. I think it can be removed but it doesn't cause any problem. So I'm leaving it as optional.

lgraf commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionPersisterProviderTest.java:
```diff
@@ -508,6 +513,82 @@ public void testExpiredSessions() {
+            });
+
+            inComittedTransaction(session -> {
+                // Persist online session
```
When it is not necessary, i will remove it.

lgraf commented:
@martin-kanis Thank you for the review. I removed the online session part from the test and rebased on main. I put the change in an extra commit, so that it is easier to review. If you are ok with the change, just let me know and i will squash the commits together before merging.

martin-kanis commented:
@lgraf Thank you! You can squash the commits.

lgraf commented:
@martin-kanis Done, thank you again for your time!

hmlnarik reviewed (approved):
Approving per @martin-kanis 's approval
